### DISCUSSION TITLE
Automations

### DISCUSSION THREAD

**Discussion by robmarkcole** [robmarkcole/HASS-Deepstack#163] [Show and tell]
For posting automations

**Comment by koldogut** [comment]
#150 
[https://github.com/robmarkcole/HASS-Deepstack-object/issues/150#issuecomment-750035833](url)

havent advanced much in the last weeks, I faced some weird troubles with the shell command (working in unix shell but not from home assistant) and stopped for a while. This command works perfectly in the shell with a bunch of files:

`'cat /config/snapshots/CAM_1*.jpg | ffmpeg -f image2pipe -r 3 -vcodec mjpeg -i - -vcodec libx264 /config/snapshots/CAM_1.mp4'`

@jodur May be you can go further on this and give me some advice to run it from HA.

**Reply by jodur** [reply]
I will give i a try, i was already looking into the ffmpeg documentation and found a [async libffmpeg](https://github.com/home-assistant-libs/ha-ffmpeg) for homeassistant that maybe could be used for this purpose. I have to dive into this and play arround. Will provide some feeback after have been playing with it.

**Reply by koldogut** [reply]
Perfect, I have 7 cameras and a dual Xeon with 24 cores (48 virtual) to mess around if you need help with the testing

**Reply by jodur** [reply]
@koldogut , i found this [https://community.home-assistant.io/t/animated-gifs-from-motion-detection-stills-for-ha/139668/7](https://community.home-assistant.io/t/animated-gifs-from-motion-detection-stills-for-ha/139668/7)

This would be  probable be a easier way, by using an animated GIF instead of mp4. Python code instead of shell scripts

**Reply by koldogut** [reply]
Probably, but mind mp4 video can be paused on a scene and slow-motioned if needed when sent to a mobile phone whereas gif can't (I pressume). In case of a robbery you can reach the video stored in the NVR for sure and gif could be easier to manage ... unless robbers take NVR too ...

**Reply by jodur** [reply]
Did write some code in python for testing the concept. With a few lines of code i now can select "BeginTimestamp" and a "Endtimestamp" and convert images within the range to a  file <latestmotion.gif > that is placed in the config/www folder. With a camera platform generic i am able to display the <lastestmotion.gif>. I first tried a picture card in lovelace, but that didn't show the updated gif, because of the browser cache.  The next level is to integrate this within homeassistant.  There are several steps to take into account
1 save timestamp at begin of Repeated image scan
2 save timestamp at end of repeated image scan
3 call the service with timestamps as parameters

1 and 2 could be realized with the new variables options in automations
3 could be solved in different ways:

3.1 Python scripts from within homeassistant is **a no go**, because i need to import external python libraries.
3.2 Use appdeamon or Pyscript, but these have a to long learncurve for me now
3.3 Use MQTT to call/trigger the service with the timestamps as payload. The service could be running on a different computer or on a other container.

For now i will proceed with 3.3. The advantage of this is that it could be fully tested and debugged on my laptop without touching my running homeassistant. When it is running well, it could be easily deployed as a Hassio addon container.

@robmarkcole  any other suggestions for this?

**Comment by robmarkcole** [comment]
@jodur 3.1 python scripts is always an option - you can safely create a virtual environment using venv and run any necessary scripts in it. This is natural suggestion from me since I am python dev

**Comment by jodur** [comment]
@koldogut and @robmarkcole if did find some time between diner preperations, family visits etc during x-mas days. I have released a custom components for homeassistant to deal with the snapshotfiles and create a gif or mp4 for the selected files. Please have a look at [https://github.com/jodur/snaptogif](https://github.com/jodur/snaptogif) and see if it usefull for your case. I am using it succesfully with the generated GIF's. I start the object recognition on the motion detection of my camera. I record the begin timestamp of motion signal and the endtime when no motion is detected any more. If the imageprocessing unit recognize objects during this time, snapshots are created. At the end of the motion i provide the `snaptogif` service with the begin and endtimestamp, and then a GIF or MP4 is created. The resulting output file is dispay using a generic camera  named  camera_motion_placexx  and is displaying the latest catched frames  with recognized object in a loop. I use it for 2 camera's now and i am happy with the results. I noticed that object recognition is more difficult in the evening and i  have a less reliable object detection. then during daylight where i have a 100% score. Maybe the model for object reconistion is only "teached" with daylight images? I will add a second service soon, that is responsible for removing or archiving the snapshots. I decided to spilt this up with a seperate service so, that we have more flexibillty what to do with the original snapshot. A scenario could be for example running a automation that is creating a mp4 movie for al snapshot, every hour, and remove the original snapshot after 1 day/week/etc..

**Reply by koldogut** [reply]
I've started with the testing, great job by the way. Unfortunately I'm facing this error while calling the service:

```
Logger: homeassistant.components.websocket_api.http.connection
Source: custom_components/snaptogif/__init__.py:59
Integration: Home Assistant WebSocket API (documentation, issues)
First occurred: 10:17:53 (5 occurrences)
Last logged: 10:20:03

[23084904807152] [Errno 2] No such file or directory: 'config/snapshots'
[23084904807152] [Errno 2] No such file or directory: 'config/snapshots/'
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/websocket_api/commands.py", line 135, in handle_call_service
    await hass.services.async_call(
  File "/usr/src/homeassistant/homeassistant/core.py", line 1445, in async_call
    task.result()
  File "/usr/src/homeassistant/homeassistant/core.py", line 1484, in _execute_service
    await self._hass.async_add_executor_job(handler.job.target, service_call)
  File "/usr/local/lib/python3.8/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/config/custom_components/snaptogif/__init__.py", line 59, in SnapToGIF
    files=os.listdir(folder)
FileNotFoundError: [Errno 2] No such file or directory: 'config/snapshots'
```

I've created a new issue on your repo [https://github.com/jodur/snaptogif/issues/1#issue-775005745](url)

**Reply by koldogut** [reply]
Solved 5 minutes later ... you must ensure folders include root `/` 

It's weird however .gif is pretty much bigger than .mp4 with the same amount of snapshots. I'll continue working with this as I want to test performance when running in parallel with multiple cameras.

**Reply by jodur** [reply]
Ok, thnks for the feedback, this shows me that i have to add validation for the directory paths too, will have a look at later on.'
At the moment i have not yet implemented compression for the GIF. The mp4 is using compression because it is using ffmpeg for encoding the mp4. In theory creating a GIF should be quicker and less CPU intensive, however it wil use more storage space. GIF will be less suitable for output with a lot of snapshots,so i think mp4 is also the prefered option for the future, but ialready had everything configured for gif's. Will test today if i could also use mp4 files and show them in a continues loop in homeassistant and notify messages on my phone.

**Reply by jodur** [reply]
Yesterday i released a update for the component. Services for move and delete files are now added to component.

**Comment by jodur** [comment]
I created a new custom component with several additions to the previous, including a camera entity for images
see:[https://community.home-assistant.io/t/face-and-person-detection-with-deepstack-local-and-free/92041/1088?u=jodur](https://community.home-assistant.io/t/face-and-person-detection-with-deepstack-local-and-free/92041/1088?u=jodur)
